### PR TITLE
Fix part of #2294: Migration to TextInputLayout for AdminAuth

### PR DESCRIPTION
**Pull request by anandwana001**
## Explanation

Fix part of #2294 
Fix part of #962 
Migrating for `AdminAuthActivity`

## Screenshots

| Portrait      | Landscape |
| ----------- | ----------- |
|       |        |
|    |        |
|    |         |
|    |         |

| Tablet Landscape | Tablet Portrait  |
| ----------- | ----------- |
|     |    | 
|        |   | 
| |  |
| |  |

**Video**
https://user-images.githubusercontent.com/25057618/102735721-2c5f4400-4369-11eb-90a4-4a9e8e994dfe.mp4

**Test Result**

| Espresso | Robolectric |
| ---------- | ---------- |
| Screenshot 2020-12-21 at 13 09 02 |  Screenshot 2020-12-21 at 13 03 42 | 

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [ ] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: app/src/main/java/org/oppia/android/app/databinding/EditTextBindingAdapters.java
Comment (anandwana001) [inline app/src/main/java/org/oppia/android/app/databinding/EditTextBindingAdapters.java]: this is setting the error text to the layout. 
Not able to do from the XML side, doing here. 
Will add the kdoc.

File: app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt
Comment (anandwana001) [inline app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt]: `adminAuthInputPinEditText.text` return Editablet so converting it to string.

File: app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt
Comment (anandwana001) [inline app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt]: Tested the IME action on emulator and physical device too. 

Internally Espresso using `performEditorAction` when we call `pressImeActionButton`, which works on event.

```Java
public boolean performEditorAction(int actionCode) {
        long eventTime = SystemClock.uptimeMillis();
        sendKeyEvent(new KeyEvent(eventTime, eventTime,
                KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_ENTER, 0, 0,
                KeyCharacterMap.VIRTUAL_KEYBOARD, 0,
                KeyEvent.FLAG_SOFT_KEYBOARD | KeyEvent.FLAG_KEEP_TOUCH_MODE
                | KeyEvent.FLAG_EDITOR_ACTION));
        sendKeyEvent(new KeyEvent(SystemClock.uptimeMillis(), eventTime,
                KeyEvent.ACTION_UP, KeyEvent.KEYCODE_ENTER, 0, 0,
                KeyCharacterMap.VIRTUAL_KEYBOARD, 0,
                KeyEvent.FLAG_SOFT_KEYBOARD | KeyEvent.FLAG_KEEP_TOUCH_MODE
                | KeyEvent.FLAG_EDITOR_ACTION));
        return true;
    }
```

File: app/src/main/res/layout-land/admin_auth_activity.xml
Comment (anandwana001) [inline app/src/main/res/layout-land/admin_auth_activity.xml]: Pre defined style to get the view of the box with a border.

File: app/src/main/res/layout-land/admin_auth_activity.xml
Comment (anandwana001) [inline app/src/main/res/layout-land/admin_auth_activity.xml]: the radius for the border. 
We can create a style for it which can be useable for every other place.

File: app/src/main/res/layout-land/admin_auth_activity.xml
Comment (anandwana001) [inline app/src/main/res/layout-land/admin_auth_activity.xml]: our custom binding adapter  `setError()`

**Comment by anandwana001** [comment]
Please skip the ktlint failure from the GitHub action check. This is just a draft PR to review the implementation and correctness of the view.

**Review by rt4914 — COMMENTED** [review]
The implementation works nicely. Will need design approval from Chantel before doing a full pass. Thanks.

**Comment by mschanteltc** [comment]
The input should have a white (#FFFFFF) fill color. Also what is the font size for "Incorrect Administrator PIN. Please try again." and "Administrator's PIN" when there is an error? Is it possible we can increase it?

**Comment by anandwana001** [comment]
> The input should have a white (#FFFFFF) fill color. Also what is the font size for "Incorrect Administrator PIN. Please try again." and "Administrator's PIN" when there is an error? Is it possible we can increase it?

@mschanteltc 
1. white color updated for box
2. The font size for error is currently the default style. I will create our own custom style, let me know the details. 
```XML
<style name="TextAppearance.MaterialComponents.Caption" parent="TextAppearance.AppCompat.Caption">
    <item name="fontFamily">sans-serif</item>
    <item name="android:fontFamily">sans-serif</item>
    <item name="android:textStyle">normal</item>
    <item name="android:textAllCaps">false</item>
    <item name="android:textSize">12sp</item>
    <item name="android:letterSpacing">0.0333333333</item>
  </style>
```

**Comment by BenHenning** [comment]
It appears I have nothing to act upon here yet, so will wait to be assigned. :)

**Comment by mschanteltc** [comment]
Hi Akshay. Here are the styles:

fontFamily: Roboto, Sans Serif
textStyle: normal
textAllCaps: false
textSize: 14sp (guesstimate, might change later)
letterSpacing: default

Also change the border, the text cursor, and the font color of "Administrator's PIN" to #333333.

File: app/src/main/res/values/styles.xml
Comment (anandwana001) [inline app/src/main/res/values/styles.xml:181]: @mschanteltc updated as per your suggestion. PTAL at screenshots as well.

File: app/src/main/res/values/styles.xml
Comment (anandwana001) [inline app/src/main/res/values/styles.xml:184]: will arrange in order, same as other styles.

**Comment by mschanteltc** [comment]
I have a question: is there a reason we are using this text input layout instead of the one visualized in the [mocks](https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/785546bd-b160-4d9b-b5d6-53ddc1b4a80a/)? 

The only issue I have with this current format is that when the label moves to the border of the text input container, it has a transparent background. But since the text input background is white, a portion of that text input container is transparent.

**Comment by anandwana001** [comment]
> I have a question: is there a reason we are using this text input layout instead of the one visualized in the [mocks](https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/785546bd-b160-4d9b-b5d6-53ddc1b4a80a/)?
> 
> The only issue I have with this current format is that when the label moves to the border of the text input container, it has a transparent background. But since the text input background is white, a portion of that text input container is transparent.

@mschanteltc The main reason for this migration is because the current implementation we have is a custom view, it's not something provided directly by Android. As it's a custom, it needs a lot of extra effort for modification, introducing any new feature is something which might a bit hard for new contributors and testing is hard. 

TextInputLayout solves all of these. As per your concern regarding the floating text at the top without transparent background, there are two types of it: One is the [outline](https://material.io/components/text-fields/android#outlined-text-field) one which is in this PR, second is of type [filled](https://material.io/components/text-fields/android#filled-text-field)

**Comment by mschanteltc** [comment]
Thank you for the explanation @anandwana001. In that case the one that is already implemented should suffice. Is there anything else you need from my end?

**Comment by anandwana001** [comment]
@mschanteltc just to confirm, the above screenshots are good to go?

**Comment by mschanteltc** [comment]
@anandwana001 Yes LGTM.

File: app/src/main/java/org/oppia/android/app/databinding/EditTextBindingAdapters.java
Comment (anandwana001) [reply app/src/main/java/org/oppia/android/app/databinding/EditTextBindingAdapters.java]: Done

File: app/src/main/res/layout-land/admin_auth_activity.xml
Comment (anandwana001) [reply app/src/main/res/layout-land/admin_auth_activity.xml]: done

File: app/src/main/res/layout-land/admin_auth_activity.xml
Comment (anandwana001) [reply app/src/main/res/layout-land/admin_auth_activity.xml]: done

File: app/src/main/res/layout-land/admin_auth_activity.xml
Comment (anandwana001) [reply app/src/main/res/layout-land/admin_auth_activity.xml]: created a new `TextInputLayoutBindingAdapters`

File: app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt
Comment (anandwana001) [reply app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt]: done

File: app/BUILD.bazel
Comment (anandwana001) [inline app/BUILD.bazel]: @BenHenning Added new file to BUILD.bazel file but the check is failing. 
Are there other steps which need to be done when we introduce new `BindingAdapters`

Comment (rt4914) [reply app/BUILD.bazel]: @anandwana001 The error is because of `TextInputLayout`, i am not sure if you might need to mention it in this section
`# Library for all custom data-binding adapters under org/oppia/android/app/databinding.`

Comment (anandwana001) [reply app/BUILD.bazel]: Worked, Thanks

File: app/src/main/res/values/dimens.xml
Comment (ShridharGoel) [inline app/src/main/res/values/dimens.xml]: Small change: Should be `stroke`.

File: app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt
Comment (ShridharGoel) [inline app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt]: Just a small typo here. Should be `confirmPin`.

**Review by ShridharGoel — CHANGES_REQUESTED** [review]
Rest looks good.

File: app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt
Comment (anandwana001) [reply app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt]: nice catch

File: app/src/main/res/values/dimens.xml
Comment (anandwana001) [reply app/src/main/res/values/dimens.xml]: nice catch, done

**Review by ShridharGoel — APPROVED** [review]
LGTM.

File: app/BUILD.bazel
Comment (rt4914) [inline app/BUILD.bazel]: This should be in alphabetical order.

File: app/src/main/java/org/oppia/android/app/databinding/TextInputLayoutBindingAdapters.java
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/databinding/TextInputLayoutBindingAdapters.java]: `.` at the end
`/** Binding adapter for setting an error message. */`

File: app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt]: Suggested optimisations:
```
    binding.adminAuthInputPinEditText.setOnEditorActionListener { _, actionId, event ->
      if (actionId == EditorInfo.IME_ACTION_DONE ||
        (event != null && (event.keyCode == KeyEvent.KEYCODE_ENTER))
      ) {
        binding.adminAuthSubmitButton.callOnClick()
      }
      false
    }
```

**Review by rt4914 — CHANGES_REQUESTED** [review]
Nit changes and one test failing PTAL

Screenshot 2020-12-22 at 11 57 51 AM

File: app/src/sharedTest/java/org/oppia/android/app/profile/AdminAuthActivityTest.kt
Comment (FareesHussain) [inline app/src/sharedTest/java/org/oppia/android/app/profile/AdminAuthActivityTest.kt:549]: - This work but I'm not sure if it is a good practice
```
      onView(withId(R.id.admin_auth_input_pin))
        .check(matches(hasDescendant(withText(R.string.admin_auth_incorrect))))
```

**Review by FareesHussain — CHANGES_REQUESTED** [review]
Failing test case that rajat mentioned [here](https://github.com/oppia/oppia-android/pull/2265#pullrequestreview-556843904) same for tablet

**Comment by BenHenning** [comment]
This appears to still be in draft. Is it ready for review?

File: app/src/sharedTest/java/org/oppia/android/app/profile/AdminAuthActivityTest.kt
Comment (FareesHussain) [reply app/src/sharedTest/java/org/oppia/android/app/profile/AdminAuthActivityTest.kt:549]: I'm sorry i was a bit confused the testcase works well 
This may be a problem with pressImeActionbutton as what I've tested as mentioned above hasErrorText works after the orientation change
please mark this conversation as resolved

File: app/BUILD.bazel
Comment (anandwana001) [reply app/BUILD.bazel]: done

File: app/src/main/java/org/oppia/android/app/databinding/TextInputLayoutBindingAdapters.java
Comment (anandwana001) [reply app/src/main/java/org/oppia/android/app/databinding/TextInputLayoutBindingAdapters.java]: done

File: app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt
Comment (anandwana001) [reply app/src/main/java/org/oppia/android/app/profile/AdminAuthActivityPresenter.kt]: done

**Comment by anandwana001** [comment]
> This appears to still be in draft. Is it ready for review?

yes, it is ready now. Was waiting for the First pass by Farees.

**Comment by BenHenning** [comment]
Thanks for clarifying. Going to defer this review to others--I suspect I won't have required codeowners for any of these files once the codeowners PR is merged.

**Comment by BenHenning** [comment]
Hmm I apparently do have codeowners files here, though some are incorrect. I'll need to iterate on the codeowners file tomorrow.

**Comment by BenHenning** [comment]
I no longer have code owners since I'm on holiday starting tomorrow--deferring to @rt4914.

**Comment by BenHenning** [comment]
(though as FYI the Bazel & adapter changes LGTM which would be the main files I would own in this case, I think).

**Review by rt4914 — APPROVED** [review]
LGTM, thanks. @anandwana001 Please update the branch.

**Review by FareesHussain — APPROVED** [review]

**Pull request merged**
